Source/miniwin/misc_msg.cpp

```diff
@@ -717,17 +724,29 @@ bool DemoMessage(tagMSG *lpMsg, int tick)
 	return false;
 }
 
-bool FetchMessage(tagMSG *lpMsg, int tick)
+bool GetDemoRunGameLoop()
+{
+	if (demo_message_queue.empty())
+		app_fatal("Demo queue empty");
+	demoMsg dmsg = demo_message_queue.front();
+	if (dmsg.type == DemoMsgType::Message)
+		app_fatal("Unexpected Message");
+	gfProgressToNextGameTick = dmsg.progressToNextGameTick;

```

Comment:
I actually like that we are recording the exact time of input, despite floats not being my favorite kinda data.

Reply:
I would also like to get rid of this. :laughing:

in the future I could imagine we would replace every call to `SDL_GetTick()` with a new introduced `DiabloGetTick()`.
And it would look something like:

```
int DiabloGetTick()
{
	if (gbRunGame)
		return gameTick;
	return SDL_GetTicks();
}

int gameTick;
void UpdateGameTick()
{
	if (timedemo) {
		gameTick = GetDemoMessageAsTime();
		return;
	}
	gameTick = SDL_GetTicks();
}
```

`UpdateGameTick()` would only be called at the top of the [loop](https://github.com/diasurgical/devilutionX/blob/fe034ce70920b5f4fdaa5ff123437f3669e13403/Source/diablo.cpp#L805) in `RunGameLoop`.
This has some benefits:

- We can remove specials like in `nthread_has_500ms_passed` or `nthread_UpdateProgressToNextGameTick`
- We don't have to care about new or changed calls To `DiabloGetTick()` causes issues in the replay. Cause there is only one point in time while the loop runs, regardless how often you ask the method. 😉 
- We don't have to store floats and implementation details like `gfProgressToNextGameTick`

I didn't add this to this PR cause it can result in unexpected errors for the normal play logic and I think it's also big enough to justify it's own PR. But I think we improve the timedemo also step for step and make this a great feature. it will take some time as it was before with ADL. 😉  But together I think we can achieve this. 🙂

Reply:
Agreed, I was thinking that maybe it could even help multiplayer if the game can communicate what the current tick is and that allow the client to better sync up with the server.